src/util/newconfig/ConfigConstraints.cpp

```diff
@@ -0,0 +1,137 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/newconfig/ConfigConstraints.hpp"
+
+#include "util/newconfig/Errors.hpp"
+
+#include <fmt/core.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <optional>
+#include <regex>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace util::config {
+
+std::optional<Error>
+PortConstraint::checkConstraint(Constraint::ValueType const& port) const
+{
+    uint32_t p = 0;
+    if (!checkType(port))
+        return Error{"Port must be an integer or string"};
+
+    if (std::holds_alternative<std::string>(port)) {
+        p = static_cast<uint32_t>(std::stoi(std::get<std::string>(port)));
+    } else {
+        p = static_cast<uint32_t>(std::get<int64_t>(port));
+    }
+    if (p >= portMin && p <= portMax)
+        return std::nullopt;
+    return Error{"Port does not satisfy the constraint bounds"};
+}
+
+std::optional<Error>
+ChannelNameConstraint::checkConstraint(Constraint::ValueType const& channel) const
+{
+    if (!checkType(channel))
+        return Error{"Channel name must be a string"};
+
+    if (std::ranges::any_of(channels, [&channel](std::string_view name) {
+            return std::get<std::string>(channel) == name;
+        }))
+        return std::nullopt;
+    return Error{"Channel name must be one of General, WebServer, Backend, RPC, ETL, Subscriptions, Performance"};
```

Comment:
Maybe the channels can be a member valuable, then there is just one place to update if we want to add/remove log channel.

Reply:
We should definitely avoid hardcoding the list here. If we add a new channel, this message will have to be updated too. Not very future proof as it's super easy to forget (or to not even know this error message exists)